### PR TITLE
Agones 

### PR DESCRIPTION
**Pull request by capturcus**
*before compiling this you must rebuild the docker image*
- included agones in the development docker image
- included agones in our build system
- added scripts for automating agones related tasks
- added a simple server module for communicating with agones
- added agones files for launching a single server or a 2 server fleet

### CODE REVIEW

File: server/agones.hpp [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#pragma once
+
+namespace agones {
+
+void Start();
+void Shutdown();
+void SetPlayers(int players);
+void SetPlaying();
+
+}; // agones
```
Comment (piodul) [inline server/agones.hpp]: Nit: missing endline

File: server/agones.cpp [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+#include <agones/sdk.h>
+#include <grpc++/grpc++.h>
+#include <iostream>
+#include <thread>
+
+static std::shared_ptr<agones::SDK> sdk;
+
+// send health check pings
+void DoHealth()
```
Comment (piodul) [inline server/agones.cpp]: Make it `static`

File: server/agones.cpp [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+		bool ok = sdk->Health();
+		if (!ok) {
+			std::cout << "health ping failed\n";
+		}
+		std::this_thread::sleep_for(std::chrono::seconds(2));
+	}
+}
+
+// watch GameServer Updates
+void WatchUpdates()
```
Comment (piodul) [inline server/agones.cpp]: Make it `static`

File: server/agones.cpp [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+	{
+		std::cerr << "Could not run GameServer(): " << status.error_message()
+				  << ". Exiting!\n";
+		exit(1);
+	}
+
+	std::cout << "GameServer name: " << gameserver.object_meta().name() << "\n"
+			  << std::flush;
+
+	health.join();
```
Comment (piodul) [inline server/agones.cpp]: Doesn't this hang this thread? The `health` thread runs an infinite loop.

Comment (piodul) [reply server/agones.cpp]: We should notify the `health` thread that we are stopping and that it should no longer be running. Here is a suggestion how to do it: https://stackoverflow.com/a/59756732/12851162

File: server/main.cpp [hunk 5]
```diff
@@ -187,6 +192,9 @@ int main(int argc, const char* const argv[])
 
 	std::cout << "server started\n";
 
+	if (gameState.options["agones"].as<bool>())
+		new std::thread(agones::Start);
```
Comment (piodul) [inline server/main.cpp]: 🙄

File: server/agones.cpp [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+			  << std::flush;
+	sdk->WatchGameServer([](const agones::dev::sdk::GameServer &gameserver) {
+		std::cout << "GameServer Update:\n"								   //
+				  << "\tname: " << gameserver.object_meta().name() << "\n" //
+				  << "\tstate: " << gameserver.status().state() << "\n"
+				  << std::flush;
+	});
+}
+
+void dfAgones::Start()
```
Comment (piodul) [inline server/agones.cpp]: I don't think this function should be blocking. Its name suggests that it should return after Agones SDK's initialization finishes.

I'd move everything after line:

```c++
	std::cout << "GameServer name: " << gameserver.object_meta().name() << "\n"
			  << std::flush;
```

to the beginning of the `dfAgones::Shutdown()` function.

File: server/agones.cpp [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+			  << "Getting the instance of the SDK.\n"
+			  << std::flush;
+	sdk = std::make_shared<agones::SDK>();
+
+	std::cout << "Attempting to connect...\n"
+			  << std::flush;
+	if (!sdk->Connect())
+	{
+		std::cerr << "Exiting!\n";
+		exit(1);
```
Comment (piodul) [inline server/agones.cpp]: IMO, the function `dfAgones::Start()` should return a boolean value indicating if we successfully started the server.

**Review by piodul — CHANGES_REQUESTED** [review]
(reviewed c++ code for now, didn't look at the YAMLs)

**Comment by capturcus** [comment]
idea: agones::Start initializes the agones sdk and if it all succeeds launches the separate agones threads containing health checks and server state watches.

File: server/agones.cpp (hunk 4 above)
Comment (capturcus) [reply server/agones.cpp]: resolved this according to the comment below

File: server/main.cpp (hunk 5 above)
Comment (capturcus) [reply server/main.cpp]: resolved this according to the comment below

File: server/agones.cpp (hunk 6 above)
Comment (capturcus) [reply server/agones.cpp]: resolved this according to the comment below

File: server/agones.cpp (hunk 7 above)
Comment (capturcus) [reply server/agones.cpp]: resolved this according to the comment below

File: server/agones.cpp [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+	}
+}
+
+// watch GameServer Updates
+static void WatchUpdates()
+{
+	std::cout << "Starting to watch GameServer updates...\n"
+			  << std::flush;
+	sdk->WatchGameServer([](const agones::dev::sdk::GameServer &gameserver) {
+		std::cout << "GameServer Update:\n"								   //
```
Comment (piodul) [inline server/agones.cpp]: Nit: There is an empty comment in this line, prepended by lots of whitespace characters

**Review by piodul — APPROVED** [review]

**Pull request merged**